api/src/main/java/io/minio/MinioProperties.java

```diff
@@ -62,14 +70,27 @@ private void setDevelopmentVersion() {
   private void setMinioClientJavaVersion(ClassLoader classLoader) throws IOException {
     if (classLoader != null) {
       Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
+      boolean minioManifestFound = false;
+      boolean minioVersionFound = false;
       while (resources.hasMoreElements()) {
-        Manifest manifest = new Manifest(resources.nextElement().openStream());
-        for (Object k : manifest.getMainAttributes().keySet()) {
-          String versionString = "Implementation-Version";
-          if (k.toString().equals(versionString)) {
-            version.set(manifest.getMainAttributes().getValue((Attributes.Name) k));
+        try (InputStream is = resources.nextElement().openStream()) {
+          Manifest manifest = new Manifest(is);
+          for (Map.Entry<Object, Object> entry : manifest.getMainAttributes().entrySet()) {
+            if (entry.getKey().toString().equals(META_INF_ATTRIB_IMPLEMENTATION_TITLE)
+                && entry.getValue().toString().equals(META_INF_ATTRIB_IMPLEMENTATION_TITLE_VALUE)) {
+              minioManifestFound = true;
```

Comment:
Why do we need this? or what exactly the problem in current approach?

Reply:
When minio is included with other libraries - via maven/gradle - this manifest resources will list all the manifests from the classpath - We specifically need the manifest from the minio jar - so we can get the version number correctly.

An example is added here https://github.com/vinu/minio-check (using minio with spring boot)
I added `7.1.4` of minio library, so I was expecting the User-agent to have this version, but it was like this `User-Agent: MinIO (x86_64; x86_64) minio-java/1.7.30`

This is because the current implementation lists all manifests from the classpath, and just takes the `Implementation-Version` so the version info is overridden by some other manifest.

Reply:
:+1: 

Just refactor the class to simplify everything like below
```diff
diff --git a/api/src/main/java/io/minio/MinioProperties.java b/api/src/main/java/io/minio/MinioProperties.java
index ad6f4c9b..a46430b4 100644
--- a/api/src/main/java/io/minio/MinioProperties.java
+++ b/api/src/main/java/io/minio/MinioProperties.java
@@ -20,7 +20,6 @@ import java.io.IOException;
 import java.net.URL;
 import java.util.Enumeration;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -28,49 +27,49 @@ import java.util.logging.Logger;
 /** Identifies and stores version information of minio-java package at run time. */
 enum MinioProperties {
   INSTANCE;
-
   private static final Logger LOGGER = Logger.getLogger(MinioProperties.class.getName());
-
   private final AtomicReference<String> version = new AtomicReference<>(null);
 
-  public String getVersion() {
-    String result = version.get();
-    if (result == null) {
-      synchronized (INSTANCE) {
-        if (version.get() == null) {
-          try {
-            ClassLoader classLoader = getClass().getClassLoader();
-            setMinioClientJavaVersion(classLoader);
-            setDevelopmentVersion();
-          } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "IOException occured", e);
-            version.set("unknown");
-          }
-          result = version.get();
-        }
-      }
+  private synchronized void setVersion() {
+    if (version.get() != null) {
+      return;
     }
-    return result;
-  }
 
-  private void setDevelopmentVersion() {
-    if (version.get() == null) {
+    ClassLoader classLoader = getClass().getClassLoader();
+    if (classLoader == null) {
       version.set("dev");
+      return;
     }
-  }
 
-  private void setMinioClientJavaVersion(ClassLoader classLoader) throws IOException {
-    if (classLoader != null) {
-      Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
-      while (resources.hasMoreElements()) {
-        Manifest manifest = new Manifest(resources.nextElement().openStream());
-        for (Object k : manifest.getMainAttributes().keySet()) {
-          String versionString = "Implementation-Version";
-          if (k.toString().equals(versionString)) {
-            version.set(manifest.getMainAttributes().getValue((Attributes.Name) k));
+    Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
+    while (resources.hasMoreElements()) {
+      try (InputStream is = resources.nextElement().openStream()) {
+        Manifest manifest = new Manifest(is);
+        boolean minioManifestFound = false;
+        for (Map.Entry<Object, Object> entry : manifest.getMainAttributes().entrySet()) {
+          String key = entry.getKey().toString();
+          String value = entry.getValue().toString();
+          if (minioManifestFound && key.equals("Implementation-Version")) {
+            version.set(value);
+            break;
           }
+
+          minioManifestFound = value.equals("minio") && key.equals("Implementation-Title");
         }
+      } catch (IOException e) {
+        LOGGER.log(Level.SEVERE, "IOException occured", e);
+        version.set("unknown");
       }
     }
   }
+
+  public String getVersion() {
+    String result = version.get();
+    if (result != null) {
+      return result;
+    }
+
+    setVersion();
+    return version.get();
+  }
 }
```

Reply:
:+1: 

Simplify everything like below
```diff
diff --git a/api/src/main/java/io/minio/MinioProperties.java b/api/src/main/java/io/minio/MinioProperties.java
index ad6f4c9b..72bc9e26 100644
--- a/api/src/main/java/io/minio/MinioProperties.java
+++ b/api/src/main/java/io/minio/MinioProperties.java
@@ -20,7 +20,6 @@ import java.io.IOException;
 import java.net.URL;
 import java.util.Enumeration;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -28,49 +27,52 @@ import java.util.logging.Logger;
 /** Identifies and stores version information of minio-java package at run time. */
 enum MinioProperties {
   INSTANCE;
-
   private static final Logger LOGGER = Logger.getLogger(MinioProperties.class.getName());
-
   private final AtomicReference<String> version = new AtomicReference<>(null);
 
-  public String getVersion() {
-    String result = version.get();
-    if (result == null) {
-      synchronized (INSTANCE) {
-        if (version.get() == null) {
-          try {
-            ClassLoader classLoader = getClass().getClassLoader();
-            setMinioClientJavaVersion(classLoader);
-            setDevelopmentVersion();
-          } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "IOException occured", e);
-            version.set("unknown");
-          }
-          result = version.get();
-        }
-      }
+  private synchronized void setVersion() {
+    if (version.get() != null) {
+      return;
     }
-    return result;
-  }
 
-  private void setDevelopmentVersion() {
-    if (version.get() == null) {
+    ClassLoader classLoader = getClass().getClassLoader();
+    if (classLoader == null) {
       version.set("dev");
+      return;
     }
-  }
 
-  private void setMinioClientJavaVersion(ClassLoader classLoader) throws IOException {
-    if (classLoader != null) {
-      Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
-      while (resources.hasMoreElements()) {
-        Manifest manifest = new Manifest(resources.nextElement().openStream());
-        for (Object k : manifest.getMainAttributes().keySet()) {
-          String versionString = "Implementation-Version";
-          if (k.toString().equals(versionString)) {
-            version.set(manifest.getMainAttributes().getValue((Attributes.Name) k));
+    Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
+    while (resources.hasMoreElements()) {
+      try (InputStream is = resources.nextElement().openStream()) {
+        Manifest manifest = new Manifest(is);
+        boolean minioManifestFound = false;
+        for (Map.Entry<Object, Object> entry : manifest.getMainAttributes().entrySet()) {
+          String key = entry.getKey().toString();
+          String value = entry.getValue().toString();
+          if (minioManifestFound && key.equals("Implementation-Version")) {
+            version.set(value);
+            return;
           }
+
+          minioManifestFound = value.equals("minio") && key.equals("Implementation-Title");
         }
+      } catch (IOException e) {
+        LOGGER.log(Level.SEVERE, "IOException occured", e);
+        version.set("unknown");
+        return;
       }
     }
+
+    version.set("dev");
+  }
+
+  public String getVersion() {
+    String result = version.get();
+    if (result != null) {
+      return result;
+    }
+
+    setVersion();
+    return version.get();
   }
 }
```

Reply:
More simpler and error free
```diff
diff --git a/api/src/main/java/io/minio/MinioProperties.java b/api/src/main/java/io/minio/MinioProperties.java
index ad6f4c9b..d10eedda 100644
--- a/api/src/main/java/io/minio/MinioProperties.java
+++ b/api/src/main/java/io/minio/MinioProperties.java
@@ -17,10 +17,11 @@
 package io.minio;
 
 import java.io.IOException;
+import java.io.InputStream;
 import java.net.URL;
 import java.util.Enumeration;
+import java.util.Map;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -28,49 +29,53 @@ import java.util.logging.Logger;
 /** Identifies and stores version information of minio-java package at run time. */
 enum MinioProperties {
   INSTANCE;
-
   private static final Logger LOGGER = Logger.getLogger(MinioProperties.class.getName());
-
   private final AtomicReference<String> version = new AtomicReference<>(null);
 
-  public String getVersion() {
-    String result = version.get();
-    if (result == null) {
-      synchronized (INSTANCE) {
-        if (version.get() == null) {
-          try {
-            ClassLoader classLoader = getClass().getClassLoader();
-            setMinioClientJavaVersion(classLoader);
-            setDevelopmentVersion();
-          } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "IOException occured", e);
-            version.set("unknown");
-          }
-          result = version.get();
-        }
-      }
+  private synchronized void setVersion() {
+    if (version.get() != null) {
+      return;
     }
-    return result;
-  }
 
-  private void setDevelopmentVersion() {
-    if (version.get() == null) {
-      version.set("dev");
+    version.set("dev");
+
+    ClassLoader classLoader = getClass().getClassLoader();
+    if (classLoader == null) {
+      return;
     }
-  }
 
-  private void setMinioClientJavaVersion(ClassLoader classLoader) throws IOException {
-    if (classLoader != null) {
+    try {
       Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
       while (resources.hasMoreElements()) {
-        Manifest manifest = new Manifest(resources.nextElement().openStream());
-        for (Object k : manifest.getMainAttributes().keySet()) {
-          String versionString = "Implementation-Version";
-          if (k.toString().equals(versionString)) {
-            version.set(manifest.getMainAttributes().getValue((Attributes.Name) k));
+        try (InputStream is = resources.nextElement().openStream()) {
+          Manifest manifest = new Manifest(is);
+          boolean minioManifestFound = false;
+          for (Map.Entry<Object, Object> entry : manifest.getMainAttributes().entrySet()) {
+            String key = entry.getKey().toString();
+            String value = entry.getValue().toString();
+            if (minioManifestFound && key.equals("Implementation-Version")) {
+              version.set(value);
+              return;
+            }
+
+            minioManifestFound = value.equals("minio") && key.equals("Implementation-Title");
           }
         }
       }
+    } catch (IOException e) {
+      LOGGER.log(Level.SEVERE, "IOException occured", e);
+      version.set("unknown");
+      return;
+    }
+  }
+
+  public String getVersion() {
+    String result = version.get();
+    if (result != null) {
+      return result;
     }
+
+    setVersion();
+    return version.get();
   }
 }
```

Reply:
Please check the latest push, hope that cleans things :)